### PR TITLE
Fixed TaskGroup and CancelScope exit issues on asyncio

### LINKED ISSUE #695: 100% CPU load after cancel
### Things to check first

- [X] I have searched the existing issues and didn't find my bug already reported there

- [X] I have checked that my bug is still present in the latest release

### AnyIO version

4.3.0

### Python version

3.9.2, 3.12.1

### What happened?

After Ctrl+C the program uses 100% CPU.
Looks like the problem in (since `call_soon` works without problems):
https://github.com/agronholm/anyio/blob/e0529a31e819064fcf1eaf7df55d1712a30bed6b/src/anyio/_backends/_asyncio.py#L231
https://github.com/python/cpython/blob/72dbea28cd3fce6fc457aaec2107a8e453073297/Lib/asyncio/base_events.py#L871

### How can we reproduce the bug?

```python
#!/usr/bin/env python3
from anyio import CancelScope, create_task_group, run, sleep


async def shield_task() -> None:
    with CancelScope(shield=True):
        await sleep(60)


async def task() -> None:
    async with create_task_group() as tg:
        tg.start_soon(shield_task)


async def main() -> None:
    async with create_task_group() as tg:
        tg.start_soon(task)


if __name__ == '__main__':
    run(main)
```
Ctrl+C

### LINKED ISSUE #698: Different cancel scope behaviour on asyncio vs Trio
### Things to check first

- [X] I have searched the existing issues and didn't find my bug already reported there

- [X] I have checked that my bug is still present in the latest release

### AnyIO version

4.3.0

### Python version

3.10.13; 3.12.1

### What happened?

Run a cancel scope within another cancel scope. Cancel the inner one then `await` something so that a cancellation exception is raised, but during handling also cancel the outer one (but do not call any more unshielded async functions before the end of the inner block). With the asyncio backend, the inner block will swallow the cancellation exception, allowing the outer block to continue running (until the next unshielded `await` in it), while the Trio backend will propagate the exception right out to the end of the outer block

By the way, the Trio docs give the impression that it is meant to work the same way as the asyncio backend does. (See discussion on Trio Discourse: [Do `Cancelled` exceptions “know” which block they belong to?](https://trio.discourse.group/t/do-cancelled-exceptions-know-which-block-they-belong-to/508))

### How can we reproduce the bug?

```python
import anyio

async def cancel_check():
    with anyio.CancelScope() as outer_scope:
        with anyio.CancelScope() as inner_scope:
            await anyio.sleep(0.1)
            inner_scope.cancel()
            print("inner cancelled")
            try:
                await anyio.sleep(0.1)
            finally:
                outer_scope.cancel()
                print("outer cancelled")
            print("should not be here inner")
        print("should not be here outer")
    print("all done")

anyio.run(cancel_check, backend="trio")
```

Output with Trio backend:

```
inner cancelled
outer cancelled
all done
```
Output with asyncio backend:

```
inner cancelled
outer cancelled
should not be here outer
all done
```

### PR DESCRIPTION
**Pull request by agronholm**
## Changes

This fixes both cancel scope exit behavior on asyncio where an outer cancel scope has been cancelled, and constant unfruitful cancellation attempts of a task that is waiting on `TaskGroup__aexit__()`.

Fixes #695. 
Fixes #698.

## Checklist

If this is a user-facing code change, like a bugfix or a new feature, please ensure that
you've fulfilled the following conditions (where applicable):

- [X] You've added tests (in `tests/`) added which would fail without your patch
- [ ] You've updated the documentation (in `docs/`, in case of behavior changes or new
features)
- [X] You've added a new changelog entry (in `docs/versionhistory.rst`).

If this is a trivial change, like a typo fix or a code reformatting, then you can ignore
these instructions.

### Updating the changelog

If there are no entries after the last release, use `**UNRELEASED**` as the version.
If, say, your patch fixes issue #123, the entry should look like this:

`* Fix big bad boo-boo in task groups (#123
<https://github.com/agronholm/anyio/issues/123>_; PR by @yourgithubaccount)`

If there's no issue linked, just link to your pull request instead by updating the
changelog after you've created the PR.

### CODE REVIEW

**Comment by agronholm** [comment]
My takeaways here:

- Semantics should match asyncio vs Trio, and in the current state of this PR, they don't
- The cancellation message: I'd prefer it to include the ID of the cancel scope where it originated, for easier debugging, but maybe we shouldn't use that ID to decide whether to swallow the exception in `CancelScope.__exit__()`

**Comment by agronholm** [comment]
What I have some difficulties understanding is this:
```py
async def test_cancelled_raises_beyond_origin() -> None:
    with CancelScope() as outer_scope:
        with CancelScope() as inner_scope:
            inner_scope.cancel()
            try:
                await checkpoint()
            finally:
                outer_scope.cancel()

            pytest.fail("checkpoint should have raised")

    assert not inner_scope.cancelled_caught
    assert outer_scope.cancelled_caught
```
By what logic should `outer_scope` set the `cancelled_caught` flag? This is what Trio's documentation says about it:
> Records whether this scope caught a [Cancelled](https://trio.readthedocs.io/en/stable/reference-core.html#trio.Cancelled) exception. This requires two things: (1) the with block exited with a [Cancelled](https://trio.readthedocs.io/en/stable/reference-core.html#trio.Cancelled) exception, and (2) this scope is the one that was responsible for triggering this [Cancelled](https://trio.readthedocs.io/en/stable/reference-core.html#trio.Cancelled) exception.

Here, the inner scope should not have it set because it doesn't catch the `CancelledError`, but the outer scope does. However, the outer scope did not trigger the `CancelledError` – it was the inner scope that did so. The outer scope was explicitly cancelled, but the `CancelledError` was passed through by the inner scope, so it doesn't belong to the outer scope.

Perhaps the logic is that, given that both cancel scopes were cancelled, it's the outer scope that's wholly responsible for the cancellation? And if the inner scope was shielded, it would be doing its own cancellation? But in that case, what would be the logic for `outer_scope.cancelled_caught` being `False`, if `outer_scope` was explicitly cancelled and `inner_scope` would raise a new `CancelledError` on exit?

I made two separate versions of this test case for clarity, and added comments reflecting how I think Trio's logic goes:
```py
async def test_cancelled_raises_beyond_origin_unshielded() -> None:
    with CancelScope() as outer_scope:
        with CancelScope() as inner_scope:
            inner_scope.cancel()
            try:
                await checkpoint()
            finally:
                outer_scope.cancel()

            pytest.fail("checkpoint should have raised")

    # Here, the outer scope is responsible for the cancellation, so the inner scope
    # won't catch the cancellation exception, but the outer scope will
    assert not inner_scope.cancelled_caught
    assert outer_scope.cancelled_caught


async def test_cancelled_raises_beyond_origin_shielded() -> None:
    with CancelScope() as outer_scope:
        with CancelScope(shield=True) as inner_scope:
            inner_scope.cancel()
            try:
                await checkpoint()
            finally:
                outer_scope.cancel()

            pytest.fail("checkpoint should have raised")

    # Here, the inner scope is the one responsible for cancellation, and given that the
    # outer scope was also cancelled, it is not considered to have "caught" the
    # cancellation, even though it swallows it, because the inner scope triggered it
    assert inner_scope.cancelled_caught
    assert not outer_scope.cancelled_caught
```
Could you validate or correct my assumptions here?

**Comment by agronholm** [comment]
After a day of debugging, I realized that I had gotten one thing wrong with my asyncio cancel scope fix: apparently when exiting a cancel scope, it should only reraise a cancellation exception if the inner scope was not shielded. I don't know how I could've deduced that from anything...

**Comment by agronholm** [comment]
It would seem that my solution of raising a new `CancelledError` was wrong to begin with. If an unshielded Trio cancel scope is exited with a cancellation error and an eligible enclosing scope is in a cancelled state, it just passes through the same exception instance.

**Comment by agronholm** [comment]
I've refactored the asyncio `CancelScope` implementation to resemble the Trio implementation much more closely now, hopefully making them easier to compare. I also fixed another bug that wasn't tested for before – empty task groups not yielding on exit on asyncio if there were no child tasks.

**Comment by Zac-HD** [comment]
I've still got a long review backlog after vacation + illness, sorry.  Skimming this briefly, it might be easier to review if we separated out the drop-py-38 changes from the taskgroup and cancelscope changes.

**Comment by agronholm** [comment]
> I've still got a long review backlog after vacation + illness, sorry. Skimming this briefly, it might be easier to review if we separated out the drop-py-38 changes from the taskgroup and cancelscope changes.

@gschaffner is reviewing this now, and I trust his insights in the matter. I wasn't going to drop 3.8 support, but it was a hard requirement to fix these issues.

File: src/anyio/_backends/_asyncio.py [hunk 1]
```diff
@@ -663,38 +696,43 @@ async def __aexit__(
...
+                    while self._tasks:
+                        try:
+                            await asyncio.wait(self._tasks)
+                        except CancelledError as exc:
+                            # Shield the scope against further cancellation attempts,
+                            # as they're not productive (#695)
+                            wait_scope.shield = True
+                            self.cancel_scope.cancel()
+                            if exc_val is None:
+                                exc_val = exc
```
[the remark this answers is not in the document]
Comment (agronholm) [reply src/anyio/_backends/_asyncio.py:717]: I identified one uncancellation issue: that the cancellation counter is incremented for every `task.cancel()` - regardless of the task to be cancelled, and then subsequently used to cancel the *host task*. The counter should only be incremented when the *host task* is cancelled. I've added the fix and a test that failed without that fix.

[the remark this answers is not in the document]
Comment (agronholm) [reply src/anyio/_backends/_asyncio.py:717]: Now that I think about this more, I believe the behavior is still incorrect if there are nested task groups with their own host tasks. :face_exhaling:

[the remark this answers is not in the document]
Comment (agronholm) [reply src/anyio/_backends/_asyncio.py:717]: These two tests will not both pass without some deeper changes:
```py
    async def test_cancel_counter_nested_scopes(self) -> None:
        with CancelScope() as root_scope:
            with CancelScope():
                root_scope.cancel()
                await checkpoint()

        assert not cast(asyncio.Task, asyncio.current_task()).cancelling()

    async def test_uncancel_after_taskgroup_cancelled(self) -> None:
        """
        Test that a cancel scope only uncancels the host task as many times as it has
        cancelled that specific task, and won't count child task cancellations towards
        that amount.

        """
        async def child_task(task_status: TaskStatus[None]) -> None:
            async with create_task_group() as tg:
                tg.start_soon(sleep, 3)
                await wait_all_tasks_blocked()
                task_status.started()

        task = asyncio.current_task()
        assert task
        with pytest.raises(CancelledError):
            async with create_task_group() as tg:
                await tg.start(child_task)
                task.cancel()

        assert task.cancelling() == 1
```

[the remark this answers is not in the document]
Comment (agronholm) [reply src/anyio/_backends/_asyncio.py:717]: I believe that the correct fix would be to always increment the specific scope's cancel counter when the scope cancels its own host task, even when the cancellation originates from an enclosing scope. And then, on exit, each scope would always decrement its own counter and uncancel its host task just as many times.

[the remark this answers is not in the document]
Comment (agronholm) [reply src/anyio/_backends/_asyncio.py:717]: I'm wondering if there's hope for restoring 3.8 compatibility after these changes.

[the remark this answers is not in the document]
Comment (agronholm) [reply src/anyio/_backends/_asyncio.py:717]: My fixes are not working; maybe I need to re-read @gschaffner 's code and see if that helps.

[the remark this answers is not in the document]
Comment (agronholm) [reply src/anyio/_backends/_asyncio.py:717]: After a good night's sleep I had another look and realized that a single line change would fix both tests.

[the remark this answers is not in the document]
Comment (agronholm) [reply src/anyio/_backends/_asyncio.py:717]: @gschaffner Ok, so here is my 2 cents on your first test: why should the task group swallow cancellations it didn't cause? That flies in the face of the whole idea of the uncancellation system. If the task is cancelled twice, it should exit with a `CancelledError` and `task.cancelling() == 2` because it was externally cancelled twice.

[the remark this answers is not in the document]
Comment (agronholm) [reply src/anyio/_backends/_asyncio.py:717]: As for the second test, that's a bit trickier. Because it's being cancelled first by the cancel scope, the first raised `CancelledError` is the one raised for the cancel scope, and the next ones are then ignored. To fix this, I think the task group should prioritize native cancellations over any cancel scope-originated ones.

[the remark this answers is not in the document]
Comment (agronholm) [reply src/anyio/_backends/_asyncio.py]: I've adapted the first test, fixing the expected `cancelling()` count, and it passes now. Hopefully this was it!

[the remark this answers is not in the document]
Comment (agronholm) [reply src/anyio/_backends/_asyncio.py]: Regarding your `test_asyncio_external_cancellation_swallowed_during_timeout()`, you're explicitly cancelling a task but the code responsible that is not uncancelling it when it's being caught, which then messes up the cancel counter.

My point is that we if we catch N cancellation errors, we can totally swallow N-1 of them so long as we make sure to reraise one of them. Imagine if we had two asyncio timeouts:
```py
async with asyncio.timeout(5), asyncio.timeout(3), create_task_group() as tg:
    tg.start_soon(refuses_cancellation_for_10_seconds)
```
This will cause the task group to catch two native cancellation errors, but it will only reraise one. When the task group is exited, I expected `task.cancelling()` to return 2 – one for each `asyncio.timeout()`. Then, each timeout CM will decrement the cancel counter which then goes to zero once they're both exited.

As for your last snippet, why do you expect `task.cancelling()` to be nonzero in the `finally:` block? The outer scope did not need to cancel the task, as the effective scope was already explicitly cancelled. As there are no yield points between exiting the inner scope and exiting the outer scope, the outer scope does not have a chance (or need to) to actually cancel the task.

Case in point – this passes:
```py
    async def test_cancelled_scope_based_checkpoint(self) -> None:
        """Regression test closely related to #698."""
        task = asyncio.current_task()
        assert task
        with CancelScope() as outer_scope:
            outer_scope.cancel()

            try:
                # The following three lines are a way to implement a checkpoint
                # function. See also https://github.com/python-trio/trio/issues/860.
                with CancelScope() as inner_scope:
                    inner_scope.cancel()
                    await checkpoint()
            finally:
                try:
                    await checkpoint()
                finally:
                    assert task.cancelling() == 1

            pytest.fail("checkpoint should have raised")

        assert not task.cancelling()
```

[the remark this answers is not in the document]
Comment (agronholm) [reply src/anyio/_backends/_asyncio.py]: Do you think the docs are wrong here? Or do you think that I have just misunderstood them.

I think the docs are correct about the cancel counter, but it doesn't really apply to us since we're not really swallowing any native cancellations anywhere because we always reraise _something_ when we catch a `CancelledError`. Only if we didn't do that would we need to think about the cancel counters.

[the remark this answers is not in the document]
Comment (agronholm) [reply src/anyio/_backends/_asyncio.py]: That `test_uncancel_after_child_task_failed()` case does look like it should not leak any cancellations, and I should fix it.

[the remark this answers is not in the document]
Comment (agronholm) [reply src/anyio/_backends/_asyncio.py]: Ok, fixed via 90812136f5032085de7f335cea0406d1b4e515ea.

File: docs/versionhistory.rst [hunk 2]
```diff
@@ -3,6 +3,22 @@ Version history
 
 This library adheres to `Semantic Versioning 2.0 <http://semver.org/>`_.
 
+**UNRELEASED**
+
+- Dropped support for Python 3.8
+  (as `#698 <https://github.com/agronholm/anyio/issues/698>`_) cannot be resolved
+  without cancel message support)
```
[the remark this answers is not in the document]
Comment (agronholm) [reply docs/versionhistory.rst]: Fixed.

File: src/anyio/_backends/_asyncio.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (agronholm) [reply src/anyio/_backends/_asyncio.py]: In `test_uncancel_laurel_like()`, the user code uncancels a cancellation it did not initiate, and this is a bug. Cancellation and uncancellation must always go hand in hand. This in fact was a problem in `CancelScope` implementation before my latest fix.

[the remark this answers is not in the document]
Comment (agronholm) [reply src/anyio/_backends/_asyncio.py]: This puts the stdlib docs in questionable light indeed:
> If end-user code is, for some reason, suppressing cancellation by catching [CancelledError](https://docs.python.org/3/library/asyncio-exceptions.html#asyncio.CancelledError), it needs to call this method to remove the cancellation state.

I feel that this might, after all, be misleading at best. I will need to take a look at the implementation of `asyncio.timeout()` to gain a better understanding of how it's used in the stdlib.

File: src/anyio/_backends/_asyncio.py [hunk 3]
```diff
@@ -496,19 +547,17 @@ def _deliver_cancellation(self, origin: CancelScope) -> bool:
...
                 waiter = task._fut_waiter  # type: ignore[attr-defined]
                 if not isinstance(waiter, asyncio.Future) or not waiter.done():
-                    origin._cancel_calls += 1
-                    if sys.version_info >= (3, 9):
-                        task.cancel(f"Cancelled by cancel scope {id(origin):x}")
-                    else:
-                        task.cancel()
+                    task.cancel(f"Cancelled by cancel scope {id(origin):x}")
+                    if task is origin._host_task:
+                        origin._cancel_calls += 1
```
[the remark this answers is not in the document]
Comment (danieldorado-cmyk) [reply src/anyio/_backends/_asyncio.py:560]: Hi @gschaffner , I made a PR with this fix we need it!

https://github.com/agronholm/anyio/pull/1093

[the remark this answers is not in the document]
Comment (gschaffner) [reply src/anyio/_backends/_asyncio.py:560]: Hi! I am not sure what you mean by "this fix". The case that this thread is about has been both fixed and included in the test suite since #790.
